### ISSUE TITLE
Standaard casing conventie voor URIs

### ISSUE THREAD

**Issue by joostfarla** [Geonovum/KP-APIs#205]
In de design rules stellen we een aantal eisen aan de [naamgeving van resources](https://geonovum.github.io/API-Designrules/#resources) mbv URI's. Hier ontbreekt een richtlijn hoe om te gaan met termen die uit meerdere woorden bestaan of termen die symbolen/leestekens bevatten. Het voorstel is om hier een nieuwe design rule voor te introduceren.

Volgens de [URI specificatie](https://tools.ietf.org/html/rfc3986) zijn URIs case-sensitive, met uitzondering van het scheme/protocol (`https://`) en het domain (`api.example.org`).

> 6.2.2.1.  Case Normalization
>
>  For all URIs, the hexadecimal digits within a percent-encoding
>  triplet (e.g., "%3a" versus "%3A") are case-insensitive and therefore
>  should be normalized to use uppercase letters for the digits A-F.>
>
>  When a URI uses components of the generic syntax, the component
>  syntax equivalence rules always apply; namely, that the scheme and
>  host are case-insensitive and therefore should be normalized to
>  lowercase.  For example, the URI <HTTP://www.EXAMPLE.com/> is
>  equivalent to <http://www.example.com/>.  The other generic syntax
>  components are assumed to be case-sensitive unless specifically
>  defined otherwise by the scheme (see Section 6.2.3).

Dat wil zeggen dat de overige elementen (path, query string en fragment) volgens de spec case-sensitive zijn. De praktijk is echter dat webservers/frameworks hier verschillend mee omgaan. Los daarvan bestaan er vele naming conventies / best practices op dit gebied.

Mijn voorstel zou zijn:

**Path segments in "kebab-case" (ook wel spinal-case genoemd)**

Voorbeeld: `https://api.example.org/v1/ruimtelijke-plannen`
Onderbouwing: Alles lowercase voorkomt verwarring in hoofdlettergebruik of bij speciale karakters, met behoud van leesbaarheid. Voorkomt verschillen in afhandeling tussen webservers/frameworks. Consistent met SEO best practices.

**Querystring keys in "lowerCamelCase"**

Voorbeeld: `https://api.example.org/v1/ruimtelijke-plannen?bevoegdGezag=12345`
Onderbouwing: Wanneer gebruikt voor filtering of sortering is de query key gelijk aan de property key in het JSON document ([waarvan de conventie ook "lowerCamelCase" is](https://geonovum.github.io/KP-APIs/API-strategie-extensies/#api-26))

**Symbolen/leestekens (met uitzondering van een koppelteken) weglaten**

Voorbeeld:
`https://api.example.org/v1/schemas` (`'` wordt vervangen door `-`)
`https://api.example.org/v1/hotel-restaurants` (`-` blijft behouden)

Bovenstaande komt naar mijn idee het meest in de buurt van de naming conventies die globaal het meest worden toegepast.

**Comment by mevdschee** [comment]
Omdat we niet weten wat de onderliggende databron is en welke characters deze toestaan in de tabel en veldnamen (en of dit case-sensitive is of niet) zou ik zo defensief mogelijk kiezen. Volgens mij is in relationele databases "snake_case" het best ondersteund voor zowel tabel als veldnamen.

MySQL ondersteunt bijvoorbeeld:

ASCII: [0-9,a-z,A-Z$_] (basic Latin letters, digits 0-9, dollar, underscore) 

bron: https://dev.mysql.com/doc/refman/5.7/en/identifiers.html

Ik zou daarom (lowercase) "snake_case" over "kebab-case" willen voorstellen voor de tabellen en de velden, zowel in de URL, de query als in de body (JSON).

**Comment by joostfarla** [comment]
> Omdat we niet weten wat de onderliggende databron van de API is en welke characters deze toestaat in de tabel en veldnamen (en of dit case-sensitive is of niet) zou ik zo defensief mogelijk kiezen. 

Het fundamentele punt hier is dat een API ontworpen zou moeten worden vanuit het perspectief van de afnemer. Daar spelen diverse aspecten een rol, waaronder het doel (en dus de scope/context) van de API, de afname-patronen behorende bij de diverse use cases (en daarmee de granulariteit van resources) en de developer experience (DX).

Design overwegingen (waaronder naamgeving) zouden dan ook niet gebaseerd moeten zijn op onderliggende implementatie-details. Hiermee wordt mede bedoeld:
* Resources zouden geen 1-op-1 mapping (en dus ontkoppeld) moeten zijn naar het onderliggende domein- of persistentie-model. Resources dienen toegespitst te zijn op afname, niet op opslag. Los daarvan dienen deze modellen los van elkaar te kunnen evolueren. Voorbeeld: het persistentie-model kan al aangepast zijn tbv nieuwe business requirements, terwijl de API backwards-compatible moet blijven. Een ander voorbeeld: als ik migreer naar een ander soort database dient de API niet te veranderen.
* Voor de afnemer zou niet zichtbaar moeten zijn welk ontwikkelplatform- of framework en wat voor soort databases of libraries er onder de motorkap worden gebruikt, hoe in de database wordt omgegaan met naamgeving en relaties, welke ontwikkel-patronen er worden toegepast etc etc.

Sterker nog: dit is [1 van de bestaande design-rules](https://geonovum.github.io/API-Designrules/#api-53) (deze is nog summier onderbouwd maar daar wordt aan gewerkt). Voor naamgeving van JSON properties hebben we ook [een reeds bestaande design rule](https://docs.geostandaarden.nl/api/API-Strategie-ext/#api-26-define-field-names-in-in-camelcase), waarvan ik niet de intentie heb die opnieuw ter discussie te stellen.

Aan de andere kant is het wel belangrijk dat een API relatief eenvoudig te implementeren moet zijn, dus ik heb daarmee wel begrip voor de suggestie. Dit zou echter wel van ondergeschikt belang moeten zijn. Los daarvan zijn er voldoende mogelijkheden om invulling te geven aan deze ontkoppeling, zoals het gebruik van DTOs (Data Transfer Objects) en/of ORMs (Object Relational Mappings).

**Comment by mevdschee** [comment]
Bedankt voor je uitgebreide reactie. Ik zal proberen net zo uitgebreid op je punten in te gaan.

### Uitvoerbaarheid standaard

>Resources dienen toegespitst te zijn op afname, niet op opslag.

Bij de gemeente Amsterdam (waar ik als software architect voor werk) hebben we ongeveer 100 datasets die we ontsluiten. Het ontwerpen van al die APIs die aansluiten op de use-case van de afnemer is niet realistisch: veelal weten we niet wie de afnemer zijn (mogen we niet vragen bij open data) en er zijn er altijd meerdere. 

>Aan de andere kant is het wel belangrijk dat een API relatief eenvoudig te implementeren moet zijn, dus ik heb daarmee wel begrip voor de suggestie. 

Dit is niet "wel belangrijk", maar essentieel. De kosten en doorlooptijd voor het maken van een API moeten laag (kunnen) zijn, dit is onze taak (en plicht). 

### Ontkoppeld persistentie-model

Ik zal nog iets dieper ingaan op de inhoudelijke argumenten:

>Resources zouden geen 1-op-1 mapping (en dus ontkoppeld) moeten zijn van het onderliggende domein- of persistentie-model.

Hier ben ik het totaal niet mee eens. De meeste APIs (bij de gemeente en volgens mij ook bij het Kadaster) zijn gemaakt voor het (CRUD) ontsluiten van een dataset. Bovendien zijn ze gemaakt in opdracht van de uitgever en niet de consument van de API. Feedback op de redenen die je noemt:

> 1. het persistentie-model kan al aangepast zijn tbv nieuwe business requirements, terwijl de API backwards-compatible moet blijven. 

Dit werkt in de praktijk niet en leidt tot lange opzoektabellen tussen namen die gebruikt worden in de API en namen die gebruikt worden in de dataopslag. Omdat de data ook via CSV dumps (en via een databaseverbinding) wordt gebruikt door onderzoekers binnen en buiten de gemeente is deze discrepantie onwerkbaar.

> 2. als ik migreer naar een ander soort database dient de API niet te veranderen.

Dit klinkt als een wens van een developer, niet van de bussiness.

### Referentie implementatie

Mijn inziens is het heel belangrijk dat er snel een bruikbare (open source) referentie implementatie van deze standaard komt, waarmee je snel een simpele dataset kan ontsluiten voor je dataportaal. Uit dit operationele ecosysteem zijn dan ook nieuwe standaarden/extensies af te leiden voor ingewikkeldere use-cases.

Ik ben bereid om deze te maken en te publiceren onder MIT.

### Dringende oproep

We kunnen (en mogen) de gemeente (en dus de burger) niet op hoge kosten jagen met een moeilijk uitvoerbare standaard (of eentje die past bij eerder gemaakte software). Commerciele bedrijven (zoals Qonduct IT) die een standaard opstellen voor te bouwen voor overheids APIs moeten dat zo doen dat het niet veel werk voor ze oplevert om de schijn van belangenverstrengeling te voorkomen. De publicatie van een (open source) referentie implementatie zou daar goed in passen.

**Comment by dvh** [comment]
@mevdschee hetgeen jij aanstipt is eigenlijk niet de keuze voor casing conventie maar een essentiële andere discussie, namelijk met welk doel deze API strategie gemaakt wordt. Daar kunnen de meningen over verschillen (die van mij is dat je een API maakt voor de afnemer, zodat deze het zo makkelijk mogelijk kan implementeren; d.w.z. als client frameworks over het algemeen beter gedijen bij `snake_case` dan bij `kebab-case`, dit een valide argument zou zijn), maar dit is wel een andere discussie en hoort m.i. hier niet thuis.

**Comment by mevdschee** [comment]
> die van mij is dat je een API maakt voor de afnemer

De afnemer neemt een datadienst af en niet de API: dit is slechts een transportprotocol voor data die in een database zit. De inconsistentie tussen API en andere vormen van data-uitwisseling is in niemands belang (behalve de partij die de API bouwt/documenteert/onderhoudt). Als de databronnen inconsistente casing hebben dan is het beter als deze in alle transportprotocollen hetzelfde is, dan wanneer de API afwijkt.

>met welk doel deze API strategie gemaakt wordt [...]
>maar dit is wel een andere discussie en hoort m.i. hier niet thuis.

De standaard moet zo gemaakt worden dat deze eenvoudig implementeerbaar is voor de overheid voor het ontsluiten van bestaande databronnen. 

Ik vind het heel vreemd dat @joostfarla en jij het hier niet mee eens zijn. Wat mij betreft moet deze discussie inderdaad ergens anders gevoerd worden als jullie dit belang niet zien.

**Comment by dvh** [comment]
> dit is slechts een transportprotocol voor data die in een database zit

Nee, de API is een extra laag bovenop de data die in plaats van ruwe data, een 'developer-friendly' interface biedt. Voorbeeld: de BAG bevat X/Y coördinaten terwijl developers willen werken met lat/lon. Een API biedt uitkomst door deze vertaalslag te maken voor clients.

**Comment by dvh** [comment]
> De standaard moet zo gemaakt worden dat deze eenvoudig implementeerbaar is voor de overheid voor het ontsluiten van bestaande databronnen.

Nee, de standaard moet in eerste zo gemaakt worden dat iedereen dezelfde guidelines hanteert (ongeacht welke database er bij welke overheidsinstantie dan ook gebruikt wordt) en vervolgens de 'large set of unknown clients' (inderdaad, die kennen we niet, kunnen apps zijn, 'things', nederlandse developers of niet-nederlandse developers) er zo laagdrempelig mogelijk mee kan werken, zodat het ook daadwerkelijk waardevol is dat er überhaupt tijd en moeite in (overheids) API's gestoken wordt. Volgens mij is dit een van de uitgangspunten van deze hele API strategie :-)

Wellicht goed om te lezen trouwens: https://tyk.io/your-data-model-is-not-an-api/. Ik vaar niet blind op tyk.io, maar ik had het zelf kunnen schrijven.

**Comment by mevdschee** [comment]
>Voorbeeld: de BAG bevat X/Y coördinaten terwijl developers willen werken met lat/lon

Wat een onzin, de onderzoekers willen daar ook mee werken, dus het doen van deze transformatie door een API is totaal op het verkeerde niveau. Als je nu als voorbeeld van JSON ipv JDBC gaf voor web developers, dan kon ik het met je eens zijn.

> en vervolgens de 'large set of unknown clients'

Misschien kan je dat uitleggen aan @joostfarla, want hij wil de API vormgeven naar de wensen van de afnemer/use-case.

> de standaard moet in eerste zo gemaakt worden dat iedereen dezelfde guidelines hanteert  en developers er zo laagdrempelig mogelijk mee kunnen werken

Je probeert "laagdrempelig voor developers" als een tegenstelling tegen "eenvoudig te implementeren voor de overheid" te stellen, maar daar ben ik het mee oneens. Inconsistente casing met de databron is daar bovendien een goed voorbeeld van.

**Comment by mevdschee** [comment]
>Wellicht goed om te lezen trouwens: https://tyk.io/your-data-model-is-not-an-api/. Ik vaar niet blind op tyk.io, maar ik had het zelf kunnen schrijven.

Er is een groot verschil tussen het maken van een applicatie en het publiceren van een (open) dataset. Dit artikel is simpelweg niet relevant voor deze laatste use-case. Deze use-case is wel waar de gemeente (ik dus) mee te maken heeft. Ik wil niet dat de standaard zo opgesteld wordt dat deze use-case niet of slecht wordt ondersteund.

Het spijt me als ik niet duidelijk was. Ik wil niet verbieden om de concepten uit het artikel toe te passen, maar juist beiden use-cases toestaan.

**Comment by dvh** [comment]
> Daar ben ik het niet mee eens, de onderzoekers ook met lat/lon werken, dus het doen van deze transformatie door een API is op het verkeerde niveau. Als je nu als voorbeeld van JSON ipv JDBC gaf voor web developers, dan kon ik het met je eens zijn.

Dus wat jou betreft gewoon X/Y teruggeven en de client laten worstelen hoe ze dit moeten omzetten naar lat/lon?

> Misschien kan je dat uitleggen aan @joostfarla, want hij wil de API vormgeven naar de wensen van de afnemer/use-case.

Volgens mij wil @joostfarla hetzelfde als ik, namelijk de API op één consistente manier vormgeven, voor in eerste instantie de afnemer van de API en niet de aanbieder.

> Je probeert "laagdrempelig voor developers" als een tegenstelling tegen "eenvoudig te implementeren voor de overheid" te stellen, maar daar ben ik het mee oneens. Inconsistente casing met de databron is daar bovendien een goed voorbeeld van.

Geen tegenstelling, ik vind het alleen totaal niet relevant of het eenvoudig is voor 'de overheid'. Hoe zou jij dan om willen gaan met het feit dat Amsterdam `snake_case` gebruikt en Kadaster `lowerCamelCase`?

**Comment by mevdschee** [comment]
>Dus wat jou betreft gewoon X/Y teruggeven en de client laten worstelen hoe ze dit moeten omzetten naar lat/lon?

Ja, of als het belangrijk is de transformatie in de database uitvoeren met een trigger, zodat deze voor iedereen beschikbaar is.

>ik vind het alleen totaal niet relevant of het eenvoudig is voor 'de overheid'.

Dat begrijp ik niet. Kan je dat toelichten?

**Comment by dvh** [comment]
> Ja, of als het belangrijk is de transformatie in de database uitvoeren met een trigger, zodat deze voor iedereen beschikbaar is.

Pas dan gelijk de kolomnamen aan aan `lowerCamelCase`, zijn we daar ook weer uit... Dat is toch hetzelfde?

> Dat begrijp ik niet. Kan je dat toelichten?

Dat het me een worst zal wezen of het moeilijk of makkelijk te implementeren is, dat is het probleem van de aanbieder. Tuurlijk zou het nergens op slaan om dingen overbodig complex te maken; we moeten niet kostte wat kost afwijken van andere standaarden om iedereen maar met werk op te zadelen. Maar als 3 gemeenten allemaal een andere standaard hebben kun je net zo goed geen standaard opstellen.

**Comment by mevdschee** [comment]
>Pas dan gelijk de kolomnamen aan aan lowerCamelCase, zijn we daar ook weer uit... Dat is toch hetzelfde?

Helaas zijn niet alle databronnen hoofdlettergevoelig, dus dat is geen goed idee. Dat was het punt.

>Dat het me een worst zal wezen of het moeilijk of makkelijk te implementeren is

Ik denk echt dat je je vergist.

>Maar als 3 gemeenten allemaal een andere standaard hebben kun je net zo goed geen standaard opstellen.

Wat bedoel je hiermee? Daarvoor zijn we hier toch?

**Comment by dvh** [comment]
> Helaas zijn niet alle databronnen hoofdlettergevoelig, dus dat is geen goed idee. Zo begon de discussie.

Daarom dient de standaard ook los te staan van de applicatie, dus de guidelines los van de databron.

> Wat bedoel je hiermee? Daarvoor zijn we hier toch?

Dan zullen er toch altijd door 2 van de 3 gemeenten concessies gedaan moeten worden? Gemeente A gebruikt `lowerCamelCase` in de databron, Gemeente B `snake_case`. Een standaard zou dus **niet** gekoppeld moeten zijn aan achterliggende systemen.

> Er is een groot verschil tussen het maken van een applicatie en het publiceren van een (open) dataset.

Volgens mij publiceer je open datasets (digitaal) juist zodat mensen er applicaties mee kunnen maken. Enfin, zolang jij de API ziet als transportmiddel om ruwe data via http te publiceren en ik als (een client-friendly) **I**nterface om **A**pplicatie's tegenaan te **P**rogrammeren gaan we hier niet uitkomen ben ik bang.

On-topic: het lijkt me wenselijk dat alle overheids API's dezelfde URI conventies hanteren en dat er derhalve iets over wordt opgenomen in de API strategie. `kebab-case` volgt de SEO best practices en lijkt een breed gedragen manier om het probleem van ongewenste karakters in een leesbare URI's op te lossen, dus ik zou daar vooral in meegaan!

**Comment by joostfarla** [comment]
On-topic: waar het mij om gaat is dat de keuze voor naming conventies los staat van je onderliggende database schema en de naming conventies die op dat niveau gelden. Dat is nou juist het mooie: Als API's dezelfde naming conventies hanteren worden daarmee de verschillen "onder de motorkap" weg-geabstraheerd. Dit heeft vanzelfsprekend impact op de implementatie, maar naar mijn mening niet dusdanig dat deze daarmee onuitvoerbaar wordt.

Off-topic: ik merk dat we in de discussie veel bestaande design rules / principes raken (ben ik mede debet aan door hier mijn onderbouwing aan op te hangen), maar die discussie moeten we echt elders gaan voeren. Dit raakt namelijk de hele ontstaansgeschiedenis en daarmee het bestaansrecht van de API strategie en het kennisplatform. @mevdschee ik nodig je graag uit om nieuwe issues op te voeren om dit soort zaken aan de kaak te stellen. Dit is waardevolle input om de design rules zo breed mogelijk toepasbaar te maken (of om bepaalde use cases juist bewust buiten scope te plaatsen).

**Comment by mevdschee** [comment]
>Daarom dient de standaard ook los te staan van de applicatie, dus de guidelines los van de databron.

Precies en daarom moet de standaard defensief worden opgesteld, zodat deze met elke databron is uit te voeren.

>zolang jij de API ziet als transportmiddel om ruwe data via http te publiceren

Ik stelde dat dit een belangrijke use-case is voor de gemeente die niet moet worden uitgesloten.

>of om bepaalde use cases juist bewust buiten scope te plaatsen

Daar zou ik ook erg mee geholpen zijn ja.

**Comment by mevdschee** [comment]
>Enfin, zolang jij de API ziet als transportmiddel om ruwe data via http te publiceren en ik als (een client-friendly) Interface om Applicatie's tegenaan te Programmeren gaan we hier niet uitkomen ben ik bang.

Ik denk dat je de essentie hier heel goed mee hebt gevangen. Als je data ontsluit ben je in mindere mate geinteresseerd in "slimme" transformaties in de API. Met name omdat je de use-case en het belang van de kwaliteit en oorsprong niet kent. Omdat de data via meerdere protocollen wordt ontsloten (vanuit het data warehouse) wordt de data transformatie op een ander niveau opgelost.

**Comment by frankvanes** [comment]
Eens met de openingspost van @joostfarla . Ik verwacht dat je het gros aan API use cases daarmee wel afdekt. Verwacht je met het oog op gebruik van kebab-case gebruik in de URL dat het handig is om een suggestie te doen dat resource-ids ook kebab-case (zouden) moeten zijn om URL's als https://api.example.org/v1/gebouwen/GEBOUW_IN_UPPERCASE te ontmoedigen?

**Comment by joostfarla** [comment]
> Verwacht je met het oog op gebruik van kebab-case gebruik in de URL dat het handig is om een suggestie te doen dat resource-ids ook kebab-case (zouden) moeten zijn om URL's als https://api.example.org/v1/gebouwen/GEBOUW_IN_UPPERCASE te ontmoedigen?

Goed punt! Ik denk dat het lastig is om die te standaardiseren, omdat we te maken hebben met diverse bestaande standaarden voor identificatie (bijv NEN3610), waar bijvoorbeeld wel punten en underscores in kunnen voorkomen en waarbij hoofdletters vaak gangbaar zijn. Misschien moeten we daarom opnemen dat de design rule expliciet geldt voor de "statische" onderdelen van de URL? Of opnemen dat er een uitzondering geldt voor formele standaarden, zoals NEN3610?

**Comment by CathyDingemanse** [comment]
Gelukkig bestaat er sinds kort een werkgroep "API Design Visie" waar de discussie over service doelstellingen van API's en bijbehorende design principes zoals service abstraction principle, loose coupling en evolvability kan worden gevoerd!

**Comment by CathyDingemanse** [comment]
> 
> Mijn voorstel zou zijn:
> 
> **Path segments in "kebab-case" (ook wel spinal-case genoemd)**
> 
> Voorbeeld: `https://api.example.org/v1/ruimtelijke-plannen`
> Onderbouwing: Alles lowercase voorkomt verwarring in hoofdlettergebruik of bij speciale karakters, met behoud van leesbaarheid. Voorkomt verschillen in afhandeling tussen webservers/frameworks. Consistent met SEO best practices.


>
> Symbolen/leestekens (met uitzondering van een koppelteken) weglaten
>
> Voorbeeld:
> https://api.example.org/v1/schemas (' wordt weggelaten)
> https://api.example.org/v1/hotel-restaurants (- blijft behouden)


Is er een onderbouwing voor het streepje? Wij stellen voor om ook het koppelteken weg te laten (muv van de padparameters, zie bovenstaande argumentatie). 
Stel dat het koppelteken niet wordt weggelaten, wordt het streepje dan verplicht? Mag ruimtelijke-plannen ook zonder streepje? Ik vind dat wel wat ver gaan. Dat levert bij ons iig een heleboel breaking changes op. 

We vinden dat dit alleen mag gelden voor de statische componenten van de URI's. Anders kunnen we veel bestaande identificaties niet gebruiken als padparameter (denk aan UUID's en codes uit landelijke tabellen). Dit is essentieel voor de gebruikersvriendelijkheid omdat deze in allerlei registraties is opgeslagen (en gevonden kan worden door een consumer!). De regel wordt daarmee dermate beperkend dat de kosten niet opwegen tegen de baten.
